source/hp/refinement.cc

```diff
@@ -207,6 +228,182 @@ namespace hp
+      const ComparisonFunction<typename identity<Number>::type> &compare_refine,
+      const ComparisonFunction<typename identity<Number>::type>
+        &compare_coarsen)
+    {
+      AssertDimension(dof_handler.get_triangulation().n_active_cells(),
+                      criteria.size());
+      Assert((p_refine_fraction >= 0) && (p_refine_fraction <= 1),
+             dealii::GridRefinement::ExcInvalidParameterValue());
+      Assert((p_coarsen_fraction >= 0) && (p_coarsen_fraction <= 1),
+             dealii::GridRefinement::ExcInvalidParameterValue());
+
+      // 1.) First extract from the vector of indicators the ones that
+      //     correspond to cells that we locally own.
+      unsigned int   n_flags_refinement(0), n_flags_coarsening(0);
+      Vector<Number> indicators_refinement(
+        dof_handler.get_triangulation().n_active_cells());
+      Vector<Number> indicators_coarsening(
+        dof_handler.get_triangulation().n_active_cells());
+      for (const auto &cell :
+           dof_handler.get_triangulation().active_cell_iterators())
+        if (!cell->is_artificial() && cell->is_locally_owned())
+          {
+            if (cell->refine_flag_set())
+              indicators_refinement(n_flags_refinement++) =
+                criteria(cell->active_cell_index());
+            else if (cell->coarsen_flag_set())
+              indicators_coarsening(n_flags_coarsening++) =
+                criteria(cell->active_cell_index());
+          }
+      indicators_refinement.grow_or_shrink(n_flags_refinement);
+      indicators_coarsening.grow_or_shrink(n_flags_coarsening);
+
+      // 2.) Determine the number of cells for p-refinement and p-coarsening on
+      //     basis of the flagged cells.
+      //
+      // 3.) Find thresholds for p-refinment and p-coarsening on only those
+      //     cells flagged for adaptation.
+      //
+      //     For cases in which no or all cells flagged for refinement and/or
+      //     coarsening are subject to p-adaptation, we usually pick thresholds
+      //     that appeal to all or none of the cells at once. However here, we
+      //     do not know which threshold would suffice for this task because the
+      //     user could provide any comparison function. Thus if necessary, we
+      //     overwrite the user's choice with suitable functions simplying
+      //     returning 'true' and 'false' for any cell with reference wrappers.
+      //     Thus, no function object copies are stored.
+      //
+      // 4.) Perform p-adaptation with absolute thresholds.
+      Number threshold_refinement(0.), threshold_coarsening(0.);
+      auto   reference_compare_refine  = std::cref(compare_refine);
+      auto   reference_compare_coarsen = std::cref(compare_coarsen);
+      if (const parallel::distributed::Triangulation<dim, spacedim>
+            *parallel_tria = dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
```

Comment:
shouldn't this also apply to the fully distributed triangulation?

Reply:
Yes, it's the same situation that led me to the decision above. I will apply the same changes here as well.